src/include/stats/thread_level_stats_collector.h

```diff
@@ -0,0 +1,327 @@
+   * @param index_oid OID of the index that the index memory usage happens
+   * @param size number of bytes being used
+   */
+  void CollectIndexMemoryUsage(catalog::db_oid_t database_oid, catalog::index_oid_t index_oid, size_t size) {
+    for (auto &metric : metric_dispatch_[StatsEventType::INDEX_MEMORY_USAGE])
+      metric->OnIndexMemoryUsage({database_oid, index_oid}, size);
+  }
+
+  /**
+   * Collector ation on index memory free
+   * @param database_oid OID of the database that the index memory free happens
+   * @param index_oid OID of the index that the index memory free happens
+   * @param size number of bytes being freed
+   */
+  void CollectIndexMemoryFree(catalog::db_oid_t database_oid, catalog::index_oid_t index_oid, size_t size) {
+    for (auto &metric : metric_dispatch_[StatsEventType::INDEX_MEMORY_FREE])
+      metric->OnIndexMemoryFree({database_oid, index_oid}, size);
+  }
+
+  /**
+   * Collector action on index memory reclaim
+   * @param database_oid OID of the database that the index memory reclaim happens
+   * @param index_oid OID of the index that the index memory reclaim happens
+   * @param size number of bytes being reclaim
+   */
+  void CollectIndexMemoryReclaim(catalog::db_oid_t database_oid, catalog::index_oid_t index_oid, size_t size) {
+    for (auto &metric : metric_dispatch_[StatsEventType::INDEX_MEMORY_RECLAIM])
+      metric->OnIndexMemoryReclaim({database_oid, index_oid}, size);
+  }
+
+  /**
+   * Collector action when a query begins
+   */
+  void CollectQueryBegin() {
+    for (auto &metric : metric_dispatch_[StatsEventType::QUERY_BEGIN]) metric->OnQueryBegin();
+  }
+
+  /**
+   * Collector action when a query ends
+   */
+  void CollectQueryEnd() {
+    for (auto &metric : metric_dispatch_[StatsEventType::QUERY_END]) metric->OnQueryEnd();
+  }
+
+  /**
+   * Collector action for a test event
+   */
+  void CollectTestNum(int number) {
+    for (auto &metric : metric_dispatch_[StatsEventType::TEST]) metric->OnTest(number);
+  }
+
+  /**
+   * @return A vector of raw data, for each registered metric. Each piece of
+   * data is guaranteed to be safe to read and remove, and the same type of
+   * metric is guaranteed to be in the same positopn in the returned vector
+   * for different instances of Collector.
+   */
+  std::vector<std::shared_ptr<AbstractRawData>> GetDataToAggregate();
+
+ private:
+  /**
+   * Registers a Metric so that its callbacks are invoked.
+   * Use this only in the constructor.
+   * @tparam metric type of Metric to register
+   * @param types A list of event types to receive updates about.
+   */
+  template <typename metric>
+  void RegisterMetric(std::vector<StatsEventType> types) {
+    auto m = std::make_shared<metric>();
+    metrics_.push_back(m);
+    for (StatsEventType type : types) metric_dispatch_[type].push_back(m);
+  }
+
+  using MetricList = std::vector<std::shared_ptr<Metric>>;
```

Comment:
`List` could mean something else in C++ [http://www.cplusplus.com/reference/list/list/](http://www.cplusplus.com/reference/list/list/), so might be better to use something like `MetricVector`?